consensus/src/dag/dag_fetcher.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::dag::types::{CertifiedNode, DAGMessage, DAGNetworkSender, NodeMetadata};
+use aptos_consensus_types::common::Round;
+use aptos_types::{epoch_state::EpochState, validator_verifier::ValidatorVerifier};
+use serde::{Deserialize, Serialize};
+use std::{sync::Arc, time::Duration};
+use tokio::sync::mpsc::{Receiver, Sender};
+
+/// Represents a request to fetch missing dependencies for `target`, `start_round` represents
+/// the first round we care about in the DAG, `exists_bitmask` is a two dimensional bitmask represents
+/// if a node exist at [start_round + index][validator_index].
+#[derive(Serialize, Deserialize, Clone)]
+struct FetchRequest {
+    target: NodeMetadata,
+    start_round: Round,
+    exists_bitmask: Vec<Vec<bool>>,
+}
+
+/// Represents a response to FetchRequest, `certified_nodes` are indexed by [round][validator_index]
+/// It should fill in gaps from the `exists_bitmask` according to the parents from the `target_digest` node.
+#[derive(Serialize, Deserialize, Clone)]
+struct FetchResponse {
+    epoch: u64,
+    certifies_nodes: Vec<Vec<CertifiedNode>>,
+}
+
+impl FetchResponse {
+    pub fn verify(
+        self,
+        _request: &FetchRequest,
+        _validator_verifier: &ValidatorVerifier,
+    ) -> anyhow::Result<Self> {
+        // TODO: verify the integrity of message
+        Ok(self)
+    }
+}
+
+impl DAGMessage for FetchRequest {
+    fn epoch(&self) -> u64 {
+        self.target.epoch()
+    }
+}
+
+impl DAGMessage for FetchResponse {
+    fn epoch(&self) -> u64 {
+        self.epoch
+    }
+}
+
+struct DagFetcher {
+    epoch_state: Arc<EpochState>,
```

Comment:
I only passed the verifier in broadcast receiver logic. is this better?

Reply:
it's just a verifier plus epoch number, I think there's many cases we need epoch numbers to create the message so it seems easier